### PR TITLE
Create Question Answer Example Workflow

### PR DESCRIPTION
**Pull request by marko-polo-cheno**
Question Answer workflow with CoQA (Conversational QA) and GPT's Turbo and GPT-4 models.

Going to add model metadata, test suite descriptions, evaluator, remote evaluator, and tests in a followup.

### CODE REVIEW

File: examples/question_answering/README.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+The data for this example integration lives in the publicly accessible S3 bucket `s3://kolena-public-datasets`.
+
+First, ensure that the `KOLENA_TOKEN` environment variable is populated in your environment. See our
+[initialization documentation](https://docs.kolena.io/installing-kolena/#initialization) for details.
+
+This project defines two scripts that perform the following operations:
+
+1. [`seed_test_suite.py`](question_answering/seed_test_suite.py) creates the following test suites:
+
+    - `question types :: CoQA`, stratified into types of questions: `what`, `who`, `how`, `did`, `where`, `was`, `when`, `is`, `why`, and `other`
```
Comment (y27choi) [inline examples/question_answering/README.md:25]: nice stratification

File: examples/question_answering/pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+description = "Example Kolena integration for question answering"
+authors = ["Kolena Engineering <eng@kolena.io>"]
+license = "Apache-2.0"
+
+[tool.poetry.dependencies]
+python = ">=3.9,<3.10"
+kolena = ">=0.82.0,<1"
+s3fs = "^2022.7.1"
+scikit-learn = "^1.1.2"
+numba = "^0.56.0"
```
Comment (y27choi) [inline examples/question_answering/pyproject.toml]: oh do we need `numba` and `scikit-learn`?

File: examples/question_answering/question_answering/workflow.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    answer: str
+    """The human created answer for the question based on the story."""
+
+    clean_answer: str
+    """A normalized and cleaned copy of the answer"""
+
+    context: List[str]
+    """The questions and answers in the history of the conversation."""
+
+    other_answer_1: str
```
Comment (y27choi) [inline examples/question_answering/question_answering/workflow.py:61]: What are the alternative answers and how are they used in the evaluation?

Comment (marko-polo-cheno) [reply examples/question_answering/question_answering/workflow.py:61]: These are alternative answers provided by the dataset. There is the main `ground_truth` and then `ground_truth v1`, `ground_truth v2`, `ground_truth v3`.

File: examples/question_answering/question_answering/workflow.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    Inference,
+)
+
+
+@dataclass(frozen=True)
+class TestSampleMetrics(MetricsTestSample):
+    """Sample-level metrics for the Question Answering workflow."""
+
+    is_correct: bool
+    """An indication of the answer being correct or wrong."""
```
Comment (y27choi) [inline examples/question_answering/question_answering/workflow.py]: I'm guessing this is based on the threshold specified in the Evaluator Config? Which metric is used to define correctness?  Let's add this to the comment.

File: examples/question_answering/pyproject.toml (hunk 2 above)
Comment (marko-polo-cheno) [reply examples/question_answering/pyproject.toml]: oops, nope!

File: examples/question_answering/question_answering/workflow.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    """Number of tokens used in the prompt."""
+
+    inference_completion_tokens: int
+    """Number of tokens used in the answer."""
+
+    inference_total_tokens: int
+    """Number of tokens used in total."""
+
+    source: str
+    """The location of the stored ground truth data."""
```
Comment (y27choi) [inline examples/question_answering/question_answering/workflow.py]: Is it just the ground truth data or does it include the inference as well?

File: examples/question_answering/question_answering/utils.py
Comment (y27choi) [inline examples/question_answering/question_answering/utils.py:1]: Are any of these functions used anywhere?

File: examples/question_answering/question_answering/seed_test_suite.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    # Organize samples into the dictionary
+    for ts, gt in data:
+        samples_by_turn[ts.turn].append((ts, gt))
+
+    # Create test cases using the organized dictionary
+    test_cases = TestCase.init_many(
+        data=[
+            (f"conversation depth :: {number:02d} :: {DATASET}", samples_by_turn[number])
+            for number in depths
+            if len(samples_by_turn[number]) > 10
```
Comment (y27choi) [inline examples/question_answering/question_answering/seed_test_suite.py]: Are we intentionally skipping small sized test cases (i.e. > 10)? How is this number chosen?

File: examples/question_answering/question_answering/workflow.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+
+    question_answer: ClassificationLabel
+    """A label to display the question and answer with the story."""
+
+
+@dataclass(frozen=True)
+class Inference(BaseInference):
+    """Inference type for the Question Answering workflow."""
+
+    answer: str
```
Comment (y27choi) [inline examples/question_answering/question_answering/workflow.py]: What do you think of using `ClassificationLabel` for `answer` to show the answer as an annotation?

File: examples/question_answering/question_answering/workflow.py (hunk 3 above)
Comment (y27choi) [reply examples/question_answering/question_answering/workflow.py:61]: It seems like there is a sample where it's marked as not `is_correct` when it's not matched with the `answer` but with one of the `other_answer_*`s.

[example](https://app.kolena.io/demo/studio?workflow=Question+Answering&testSuiteId=1330&testCaseId=15939&models=N4IglgJiBcBsBMBWANCAxgewHYDMwHMBXAJwEMAXMbGUSGARkVgHYBfVoA&pageSize=50&overlays=N4Igtg9gJgpgNgZxALmAXwDQgIYDtcQAu2hAlhLkqiITAoQPoLZgAOcMDYMhATqQGMq6LKVwAzGLxi4BMFCJABzXhACuuKAz5rCACwUgAjmrpkKDPAgDuUw3qIBrGAE8UIAIwgsAN1IJSACMOFB0YNEwaMyYWdnlUSNp6GLYOLh5sKBJsBQi0IA&filters=result.625.1567.is_correct%3Afalse%3Avi&sortByType=test_sample_metrics&sortBy=625.1567.MEAN_METRIC&testSampleId=25101416)

**Comment by y27choi** [comment]
Done reviewing. Left some comments (mostly questions) but the results look great 👍

File: examples/question_answering/question_answering/workflow.py (hunk 3 above)
Comment (marko-polo-cheno) [reply examples/question_answering/question_answering/workflow.py:61]: The extra answers are not used in evaluation for simplicity (like a rank=1 configuration). So its expected that there are samples that have matching inferences with the other answers but are marked as incorrect.

Comment (marko-polo-cheno) [reply examples/question_answering/question_answering/workflow.py:61]: I can change the evaluation in followups with more time to save the computed metrics!

File: examples/question_answering/question_answering/workflow.py (hunk 5 above)
Comment (marko-polo-cheno) [reply examples/question_answering/question_answering/workflow.py]: line 94: The total inference token count includes the prompt and result.
line 97: This should say the inference data!

File: examples/question_answering/question_answering/utils.py
Comment (marko-polo-cheno) [reply examples/question_answering/question_answering/utils.py:1]: Yes, but I'll add the evaluator in a followup. These utils were mainly copy and pasted from the text summ example.

File: examples/question_answering/question_answering/seed_test_suite.py (hunk 6 above)
Comment (marko-polo-cheno) [reply examples/question_answering/question_answering/seed_test_suite.py]: I intentionally skipped the small sized test cases (there were only 4 samples in them) for a clean number of 20 test cases + 1 complete test case.

Comment (y27choi) [reply examples/question_answering/question_answering/seed_test_suite.py]: hmm yeah I think maybe adding that as a comment somewhere could be helpful for anyone who's going through this example.

File: examples/question_answering/question_answering/workflow.py (hunk 5 above)
Comment (y27choi) [reply examples/question_answering/question_answering/workflow.py]: Bump

**Review by y27choi — APPROVED** [review]

File: examples/question_answering/pyproject.toml [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+description = "Example Kolena integration for question answering"
+authors = ["Kolena Engineering <eng@kolena.io>"]
+license = "Apache-2.0"
+
+[tool.poetry.dependencies]
+python = ">=3.9,<3.10"
+kolena = ">=0.82.0,<1"
+s3fs = "^2022.7.1"
+evaluate = "^0.4.0"
+pyarrow = "^12.0.1"
```
Comment (gordonhart) [inline examples/question_answering/pyproject.toml]: Why is this pin necessary?

File: examples/question_answering/pyproject.toml [hunk 9]
```diff
@@ -0,0 +1,23 @@
+[tool.poetry]
+name = "question_answering"
+version = "0.1.0"
+description = "Example Kolena integration for question answering"
+authors = ["Kolena Engineering <eng@kolena.io>"]
+license = "Apache-2.0"
+
+[tool.poetry.dependencies]
+python = ">=3.9,<3.10"
```
Comment (munkyshi) [inline examples/question_answering/pyproject.toml]: Is python 3.9 the only version that works? Ideally our examples work with a wide range of python versions.

Comment (munkyshi) [inline examples/question_answering/pyproject.toml]: Is python 3.9 the only version that works? Ideally our examples work with a wide range of python versions.

File: examples/question_answering/question_answering/workflow.py (hunk 3 above)
Comment (munkyshi) [reply examples/question_answering/question_answering/workflow.py:61]: Is the plan to eventually mark it as `is_correct` when it matches one of the `other_answer_*`s?

File: examples/question_answering/pyproject.toml (hunk 8 above)
Comment (marko-polo-cheno) [reply examples/question_answering/pyproject.toml]: Without it, `poetry install` fails:
```
  • Updating pyarrow (12.0.1 -> 13.0.0): Failed
  RuntimeError
  Unable to find installation candidates for pyarrow (13.0.0)
```

Comment (gordonhart) [reply examples/question_answering/pyproject.toml]: `poetry update && poetry install` works fine for me on a fresh clone and installs the latest version:

```
$ poetry run pip freeze | rg pyarrow
pyarrow==13.0.0
```

**Comment by marko-polo-cheno** [comment]
@munkyshi yes, the plan is to eventually mark it as `is_correct` when it better matches one of the `other_answer_*`s, but this leads to a more complicated evaluator. We can discuss and reconsider in followups

File: examples/question_answering/pyproject.toml (hunk 8 above)
Comment (marko-polo-cheno) [reply examples/question_answering/pyproject.toml]: `poetry cache clear --all .` made things work, thanks Gordon!

**Review by munkyshi — APPROVED** [review]

**Pull request merged**